File: tif_customization/tif_customization/page/dispatch_detail_report/dispatch_detail_report.py
import json

import frappe
from frappe import _
from frappe.utils import add_days, cint, flt, getdate, today

from tif_customization.tif_customization.utils.supply_chain_books import (
	is_supply_chain_book_item,
)


@frappe.whitelist()
def get_dispatch_detail_data(filters=None):
	"""Dispatch detail from Delivery Notes: customer, address, warehouse, books, courier & customer amounts."""
	try:
		if isinstance(filters, str):
			filters = json.loads(filters)
		filters = filters or {}

		from_date = getdate(filters.get("from_date") or add_days(today(), -30))
		to_date = getdate(filters.get("to_date") or today())

		params = {"from_date": from_date, "to_date": to_date}
		where = [
			"dn.docstatus = 1",
			"dn.posting_date BETWEEN %(from_date)s AND %(to_date)s",
		]

		if filters.get("customer"):
			where.append("dn.customer = %(customer)s")
			params["customer"] = filters["customer"]

		if filters.get("delivery_mode"):
			where.append("dn.custom_delivery_mode = %(delivery_mode)s")
			params["delivery_mode"] = filters["delivery_mode"]

		if filters.get("courier"):
			where.append("dn.custom_courier = %(courier)s")
			params["courier"] = filters["courier"]

		if filters.get("warehouse"):
			where.append(
				"""EXISTS (
					SELECT 1 FROM `tabDelivery Note Item` dni_w
					WHERE dni_w.parent = dn.name AND dni_w.warehouse = %(warehouse)s
				)"""
			)
			params["warehouse"] = filters["warehouse"]

		if filters.get("city"):
			where.append("(addr.city = %(city)s OR dn.custom_city = %(city)s)")
			params["city"] = filters["city"]

		where_clause = " AND ".join(where)

		rows = frappe.db.sql(
			f"""
			SELECT
				dn.name AS delivery_note_no,
				dn.posting_date,
				dn.customer,
				dn.customer_name,
				dn.shipping_address_name,
				dn.shipping_address,
				dn.custom_delivery_mode,
				dn.custom_courier,
				dn.custom_courier_service,
				dn.custom_delivery_rate,
				dn.custom_courier_mode_of_payment,
				dn.custom_city,
				IFNULL(dn.is_return, 0) AS is_return,
				dn.return_against,
				COALESCE(addr.custom_area, dn.custom_area, '') AS area,
				COALESCE(addr.city, dn.custom_city, '') AS city,
				COALESCE(addr.state, '') AS province,
				COALESCE(addr.country, '') AS country,
				TRIM(BOTH ', ' FROM CONCAT_WS(', ',
					NULLIF(addr.address_line1, ''),
					NULLIF(addr.address_line2, ''),
					NULLIF(COALESCE(addr.city, dn.custom_city), ''),
					NULLIF(addr.state, ''),
					NULLIF(addr.custom_area, ''),
					NULLIF(addr.country, '')
				)) AS shipping_address_text,
				COALESCE((
					SELECT SUM(tc.amount)
					FROM `tabTransport Charges` tc
					WHERE tc.parent = dn.name
					AND tc.parenttype = 'Delivery Note'
					AND tc.parentfield = 'custom_transport_charges'
				), 0) AS transport_charges
			FROM `tabDelivery Note` dn
			LEFT JOIN `tabAddress` addr ON addr.name = dn.shipping_address_name
			WHERE {where_clause}
			ORDER BY dn.posting_date DESC, dn.name DESC
			""",
			params,
			as_dict=True,
		)

		if not rows:
			return {"rows": [], "summary": _empty_summary()}

		dn_names = [r.delivery_note_no for r in rows]
		items_by_dn = _get_items_by_delivery_note(dn_names)
		jv_courier_by_dn = _get_jv_courier_expense_by_dn(dn_names)
		jv_details_by_dn = _get_jv_courier_details_by_dn(dn_names)

		result_rows = []
		for row in rows:
			dn_name = row.delivery_note_no
			items = items_by_dn.get(dn_name, [])
			warehouses = sorted({i["warehouse"] for i in items if i.get("warehouse")})

			jv_amount = jv_courier_by_dn.get(dn_name)
			courier_payable = _courier_payable(row, jv_amount)
			total_books = sum(flt(i.get("qty", 0)) for i in items if i.get("is_book"))
			total_qty = sum(flt(i.get("qty", 0)) for i in items)
			total_cartons = sum(flt(i.get("custom_cartons", 0)) for i in items)
			books_cost = sum(flt(i.get("book_cost", 0)) for i in items)
			jv_details = jv_details_by_dn.get(dn_name, [])

			result_rows.append(
				{
					"delivery_note_no": dn_name,
					"posting_date": row.posting_date,
					"customer": row.customer,
					"customer_name": row.customer_name or row.customer,
					"shipping_address": row.shipping_address_text
					or _strip_html(row.shipping_address)
					or "",
					"city": row.city or "",
					"area": row.area or "",
					"province": row.province or "",
					"country": row.country or "",
					"warehouses": warehouses,
					"warehouses_label": ", ".join(warehouses) if warehouses else "",
					"delivery_mode": row.custom_delivery_mode or "",
					"courier": row.custom_courier or "",
					"courier_service": row.custom_courier_service or "",
					"courier_mode_of_payment": row.custom_courier_mode_of_payment or "",
					"delivery_rate": flt(row.custom_delivery_rate),
					"transport_charges": flt(row.transport_charges),
					"jv_courier_amount": flt(jv_amount) if jv_amount is not None else 0.0,
					"courier_jv_entries": jv_details,
					"courier_payable": courier_payable,
					"courier_expense_source": "jv"
					if jv_amount is not None and flt(jv_amount) > 0
					else ("delivery_rate" if (row.custom_delivery_mode or "") == "Courier" else "transport"),
					"books_cost": books_cost,
					"total_books": total_books,
					"total_qty": total_qty,
					"total_cartons": total_cartons,
					"is_return": cint(row.is_return),
					"return_against": row.return_against or "",
					"items": items,
				}
			)

		return {"rows": result_rows, "summary": _build_summary(result_rows)}
	except Exception:
		frappe.log_error(frappe.get_traceback(), "Dispatch Detail Report Error")
		return {"error": _("Failed to load dispatch detail report."), "rows": [], "summary": _empty_summary()}


@frappe.whitelist()
def get_filter_options():
	"""Filter dropdown values."""
	couriers = frappe.get_all("Courier", fields=["name"], order_by="name")
	warehouses = frappe.get_all(
		"Warehouse", filters={"is_group": 0, "disabled": 0}, fields=["name"], order_by="name"
	)
	cities = frappe.db.sql(
		"""
		SELECT DISTINCT city FROM (
			SELECT city FROM `tabAddress` WHERE city IS NOT NULL AND city != ''
			UNION
			SELECT custom_city AS city FROM `tabDelivery Note`
			WHERE custom_city IS NOT NULL AND custom_city != ''
		) t ORDER BY city
		""",
		as_dict=True,
	)
	return {
		"couriers": [c.name for c in couriers],
		"warehouses": [w.name for w in warehouses],
		"cities": [c.city for c in cities if c.city],
		"delivery_modes": ["Courier", "Transport", "By Hand"],
	}


def _get_items_by_delivery_note(dn_names):
	items_by_dn = {}
	if not dn_names:
		return items_by_dn

	items = frappe.db.sql(
		"""
		SELECT
			dni.parent,
			dni.item_code,
			dni.item_name,
			dni.qty,
			dni.stock_uom,
			dni.warehouse,
			dni.incoming_rate,
			dni.custom_cartons,
			COALESCE(
				NULLIF(dni.incoming_rate, 0),
				NULLIF(it.valuation_rate, 0),
				NULLIF(it.last_purchase_rate, 0),
				0
			) AS unit_cost
		FROM `tabDelivery Note Item` dni
		LEFT JOIN `tabItem` it ON it.name = dni.item_code
		WHERE dni.parent IN %(dn_names)s
		ORDER BY dni.parent, dni.idx
		""",
		{"dn_names": tuple(dn_names)},
		as_dict=True,
	)

	for item in items:
		parent = item.parent
		qty = flt(item.qty)
		unit_cost = flt(item.unit_cost)
		is_book = is_supply_chain_book_item(item.item_code)
		items_by_dn.setdefault(parent, []).append(
			{
				"item_code": item.item_code,
				"item_name": item.item_name,
				"qty": qty,
				"stock_uom": item.stock_uom,
				"warehouse": item.warehouse or "",
				"unit_cost": unit_cost,
				"book_cost": qty * unit_cost if is_book else 0.0,
				"line_cost": qty * unit_cost,
				"custom_cartons": flt(item.custom_cartons),
				"book_type": _book_type(item.item_name, item.item_code) if is_book else "Other",
				"is_book": is_book,
			}
		)
	return items_by_dn


def _get_jv_courier_expense_by_dn(dn_names):
	"""Actual courier expense from Journal Entry (same basis as Courier Report)."""
	if not dn_names:
		return {}

	rows = frappe.db.sql(
		"""
		SELECT
			je.cheque_no AS dn_name,
			COALESCE(SUM(jea.debit - jea.credit), 0) AS amount
		FROM `tabJournal Entry Account` jea
		INNER JOIN `tabJournal Entry` je ON je.name = jea.parent AND je.docstatus = 1
		WHERE je.cheque_no IN %(dn_names)s
		AND (jea.account LIKE %(courier_ac)s OR jea.account LIKE %(courier_exp_ac)s)
		GROUP BY je.cheque_no
		""",
		{
			"dn_names": tuple(dn_names),
			"courier_ac": "%Courier%",
			"courier_exp_ac": "%Courier Expense%",
		},
		as_dict=True,
	)
	return {row.dn_name: flt(row.amount) for row in rows}


def _get_jv_courier_details_by_dn(dn_names):
	"""Per-DN Journal Entry lines for courier charges (for expand detail)."""
	if not dn_names:
		return {}

	rows = frappe.db.sql(
		"""
		SELECT
			je.cheque_no AS dn_name,
			je.name AS journal_entry,
			je.posting_date,
			jea.account,
			COALESCE(jea.debit - jea.credit, 0) AS amount
		FROM `tabJournal Entry Account` jea
		INNER JOIN `tabJournal Entry` je ON je.name = jea.parent AND je.docstatus = 1
		WHERE je.cheque_no IN %(dn_names)s
		AND (jea.account LIKE %(courier_ac)s OR jea.account LIKE %(courier_exp_ac)s)
		AND COALESCE(jea.debit - jea.credit, 0) != 0
		ORDER BY je.posting_date, je.name
		""",
		{
			"dn_names": tuple(dn_names),
			"courier_ac": "%Courier%",
			"courier_exp_ac": "%Courier Expense%",
		},
		as_dict=True,
	)

	out = {}
	for row in rows or []:
		out.setdefault(row.dn_name, []).append(
			{
				"journal_entry": row.journal_entry,
				"posting_date": row.posting_date,
				"account": row.account or "",
				"amount": flt(row.amount),
			}
		)
	return out


def _courier_payable(row, jv_amount=None):
	mode = row.custom_delivery_mode or ""
	if mode == "Courier":
		if jv_amount is not None and flt(jv_amount) > 0:
			return flt(jv_amount)
		return flt(row.custom_delivery_rate)
	if mode == "Transport":
		return flt(row.transport_charges)
	return 0.0


def _book_type(item_name, item_code):
	name = (item_name or "").upper()
	code = (item_code or "").upper()
	if "MQH" in name or "MQH" in code:
		return "MQH"
	if "QAIDA" in name or "QAIDA" in code or "GUIDE" in name:
		return "Qaida"
	return "Other"


def _strip_html(text):
	if not text:
		return ""
	import re

	return re.sub(r"<[^>]+>", " ", text).replace("\n", ", ").strip()


def _empty_summary():
	return {
		"total_delivery_notes": 0,
		"total_books": 0,
		"total_courier_payable": 0,
		"total_books_cost": 0,
	}


def _build_summary(rows):
	outbound = [r for r in rows if not cint(r.get("is_return"))]
	returns = [r for r in rows if cint(r.get("is_return"))]
	return {
		"total_delivery_notes": len(rows),
		"total_books": flt(sum(r["total_books"] for r in outbound)),
		"total_courier_payable": flt(sum(r["courier_payable"] for r in rows)),
		"total_books_cost": flt(sum(r["books_cost"] for r in outbound)),
		"return_delivery_notes": len(returns),
	}

File: tif_customization/tif_customization/page/dispatch_detail_report/dispatch_detail_report.js
frappe.pages["dispatch-detail-report"].on_page_load = function (wrapper) {
	const $wrapper = $(wrapper);
	if ($wrapper.find(".ddr-container").length) {
		$wrapper.find(".ddr-container").closest(".layout-main-section").empty();
	}

	const page = frappe.ui.make_app_page({
		parent: wrapper,
		title: __("Dispatch Detail Report"),
		single_column: true,
	});

	new window.DispatchDetailReport(page).make();
};

if (typeof window.DispatchDetailReport === "undefined") {
	window.DispatchDetailReport = class DispatchDetailReport {
	constructor(page) {
		this.page = page;
		this.filters = {
			from_date: frappe.datetime.add_days(frappe.datetime.get_today(), -30),
			to_date: frappe.datetime.get_today(),
			customer: "",
			delivery_mode: "",
			courier: "",
			warehouse: "",
			city: "",
		};
		this.data = { rows: [], summary: {} };
		this.filter_options = {};
	}

	make() {
		this.render_layout();
		this.setup_link_fields();
		this.load_filter_options();
		this.bind_events();
		this.load_data();
	}

	render_layout() {
		this.page.main.html(`
			<div class="ddr-container" style="padding:16px;">
				<div class="filter-section" style="background:#f8f9fa;padding:16px;border-radius:8px;margin-bottom:16px;">
					<div class="row">
						<div class="col-md-2"><div id="ddr-from-date"></div></div>
						<div class="col-md-2"><div id="ddr-to-date"></div></div>
						<div class="col-md-2"><div id="ddr-customer"></div></div>
						<div class="col-md-2">
							<label>${__("Delivery Mode")}</label>
							<select id="ddr-delivery-mode" class="form-control"><option value="">${__("All")}</option></select>
						</div>
						<div class="col-md-2">
							<label>${__("Courier")}</label>
							<select id="ddr-courier" class="form-control"><option value="">${__("All")}</option></select>
						</div>
						<div class="col-md-2">
							<label>${__("Warehouse")}</label>
							<select id="ddr-warehouse" class="form-control"><option value="">${__("All")}</option></select>
						</div>
					</div>
					<div class="row" style="margin-top:10px;">
						<div class="col-md-3">
							<label>${__("City")}</label>
							<select id="ddr-city" class="form-control"><option value="">${__("All Cities")}</option></select>
						</div>
						<div class="col-md-9" style="padding-top:24px;">
							<button class="btn btn-primary btn-sm" id="ddr-apply"><i class="fa fa-filter"></i> ${__("Apply")}</button>
							<button class="btn btn-default btn-sm" id="ddr-reset">${__("Reset")}</button>
							<button class="btn btn-success btn-sm pull-right" id="ddr-export"><i class="fa fa-file-excel-o"></i> ${__("Export")}</button>
						</div>
					</div>
				</div>

				<div class="row" id="ddr-kpis" style="margin-bottom:16px;"></div>

				<div class="table-responsive">
					<table class="table table-bordered table-hover" id="ddr-table">
						<thead>
							<tr>
								<th style="width:28px;"></th>
								<th>${__("Delivery Note")}</th>
								<th>${__("Date")}</th>
								<th>${__("Customer")}</th>
								<th>${__("Ship To Address")}</th>
								<th>${__("Warehouse")}</th>
								<th>${__("Mode")}</th>
								<th>${__("Courier")}</th>
								<th class="text-right">${__("Books")}</th>
								<th class="text-right">${__("Courier Payable")}</th>
								<th class="text-right">${__("Customer Amount")}</th>
								<th class="text-right">${__("Amount To Receive")}</th>
							</tr>
						</thead>
						<tbody id="ddr-tbody">
							<tr><td colspan="12" class="text-center text-muted">${__("Loading...")}</td></tr>
						</tbody>
					</table>
				</div>
			</div>
		`);

		this.from_date_control = frappe.ui.form.make_control({
			parent: this.page.main.find("#ddr-from-date"),
			df: { fieldtype: "Date", label: __("From Date"), fieldname: "from_date" },
			render_input: true,
		});
		this.to_date_control = frappe.ui.form.make_control({
			parent: this.page.main.find("#ddr-to-date"),
			df: { fieldtype: "Date", label: __("To Date"), fieldname: "to_date" },
			render_input: true,
		});
		this.from_date_control.set_value(this.filters.from_date);
		this.to_date_control.set_value(this.filters.to_date);
	}

	setup_link_fields() {
		this.customer_control = frappe.ui.form.make_control({
			parent: this.page.main.find("#ddr-customer"),
			df: {
				fieldtype: "Link",
				options: "Customer",
				label: __("Customer"),
				fieldname: "customer",
			},
			render_input: true,
		});
	}

	load_filter_options() {
		frappe.call({
			method:
				"tif_customization.tif_customization.page.dispatch_detail_report.dispatch_detail_report.get_filter_options",
			callback: (r) => {
				if (!r.message) return;
				this.filter_options = r.message;
				this.populate_select("#ddr-delivery-mode", r.message.delivery_modes || []);
				this.populate_select("#ddr-courier", r.message.couriers || []);
				this.populate_select("#ddr-warehouse", r.message.warehouses || []);
				this.populate_select("#ddr-city", r.message.cities || []);
			},
		});
	}

	populate_select(selector, values) {
		const $el = this.page.main.find(selector);
		values.forEach((v) => $el.append(`<option value="${frappe.utils.escape_html(v)}">${frappe.utils.escape_html(v)}</option>`));
	}

	bind_events() {
		this.page.main.find("#ddr-apply").on("click", () => this.apply_filters());
		this.page.main.find("#ddr-reset").on("click", () => this.reset_filters());
		this.page.main.find("#ddr-export").on("click", () => this.export_csv());
		this.page.main.on("click", ".ddr-toggle", (e) => {
			const dn = $(e.currentTarget).data("dn");
			this.page.main.find(`#ddr-detail-${CSS.escape(dn)}`).toggle();
			$(e.currentTarget).find("i").toggleClass("fa-chevron-right fa-chevron-down");
		});
	}

	apply_filters() {
		this.filters.from_date = this.from_date_control.get_value();
		this.filters.to_date = this.to_date_control.get_value();
		this.filters.customer = this.customer_control.get_value() || "";
		this.filters.delivery_mode = this.page.main.find("#ddr-delivery-mode").val() || "";
		this.filters.courier = this.page.main.find("#ddr-courier").val() || "";
		this.filters.warehouse = this.page.main.find("#ddr-warehouse").val() || "";
		this.filters.city = this.page.main.find("#ddr-city").val() || "";
		this.load_data();
	}

	reset_filters() {
		this.filters = {
			from_date: frappe.datetime.add_days(frappe.datetime.get_today(), -30),
			to_date: frappe.datetime.get_today(),
			customer: "",
			delivery_mode: "",
			courier: "",
			warehouse: "",
			city: "",
		};
		this.from_date_control.set_value(this.filters.from_date);
		this.to_date_control.set_value(this.filters.to_date);
		this.customer_control.set_value("");
		this.page.main.find("#ddr-delivery-mode, #ddr-courier, #ddr-warehouse, #ddr-city").val("");
		this.load_data();
	}

	load_data() {
		const $tbody = this.page.main.find("#ddr-tbody");
		$tbody.html(`<tr><td colspan="12" class="text-center text-muted">${__("Loading...")}</td></tr>`);

		frappe.call({
			method:
				"tif_customization.tif_customization.page.dispatch_detail_report.dispatch_detail_report.get_dispatch_detail_data",
			args: { filters: this.filters },
			freeze: true,
			freeze_message: __("Loading dispatch details..."),
			callback: (r) => {
				if (r.message?.error) {
					frappe.msgprint({ title: __("Error"), message: r.message.error, indicator: "red" });
					return;
				}
				this.data = r.message || { rows: [], summary: {} };
				this.render_kpis();
				this.render_table();
			},
		});
	}

	render_kpis() {
		const s = this.data.summary || {};
		const cards = [
			{ label: __("Delivery Notes"), value: s.total_delivery_notes || 0, color: "#0F62FE" },
			{ label: __("Total Books"), value: this.fmt_num(s.total_books), color: "#198038" },
			{ label: __("Courier Payable"), value: this.fmt_currency(s.total_courier_payable), color: "#DA1E28" },
			{ label: __("Customer Amount"), value: this.fmt_currency(s.total_customer_amount), color: "#8A3FFC" },
			{ label: __("Amount To Receive"), value: this.fmt_currency(s.total_amount_to_receive), color: "#FF832B" },
		];

		this.page.main.find("#ddr-kpis").html(
			cards
				.map(
					(c) => `
				<div class="col-md-2 col-sm-4 col-xs-6" style="margin-bottom:8px;">
					<div style="background:#fff;border-left:4px solid ${c.color};padding:12px 14px;border-radius:6px;box-shadow:0 1px 3px rgba(0,0,0,.08);">
						<div style="font-size:11px;color:#666;text-transform:uppercase;">${c.label}</div>
						<div style="font-size:20px;font-weight:600;margin-top:4px;">${c.value}</div>
					</div>
				</div>`
				)
				.join("")
		);
	}

	render_table() {
		const rows = this.data.rows || [];
		const $tbody = this.page.main.find("#ddr-tbody");

		if (!rows.length) {
			$tbody.html(`<tr><td colspan="12" class="text-center text-muted">${__("No records found")}</td></tr>`);
			return;
		}

		$tbody.empty();
		rows.forEach((row) => {
			const dn = row.delivery_note_no;
			const dn_link = `<a href="/app/delivery-note/${encodeURIComponent(dn)}">${frappe.utils.escape_html(dn)}</a>`;

			$tbody.append(`
				<tr class="ddr-main-row">
					<td><button class="btn btn-xs btn-default ddr-toggle" data-dn="${frappe.utils.escape_html(dn)}"><i class="fa fa-chevron-right"></i></button></td>
					<td>${dn_link}</td>
					<td>${frappe.datetime.str_to_user(row.posting_date)}</td>
					<td>${frappe.utils.escape_html(row.customer_name || "")}</td>
					<td style="max-width:220px;white-space:normal;font-size:12px;">${frappe.utils.escape_html(row.shipping_address || "")}</td>
					<td style="font-size:12px;">${frappe.utils.escape_html(row.warehouses_label || "")}</td>
					<td>${frappe.utils.escape_html(row.delivery_mode || "")}</td>
					<td style="font-size:12px;">${frappe.utils.escape_html([row.courier, row.courier_service].filter(Boolean).join(" / "))}</td>
					<td class="text-right">${this.fmt_num(row.total_books)}</td>
					<td class="text-right">${this.fmt_currency(row.courier_payable)}</td>
					<td class="text-right">${this.fmt_currency(row.customer_amount)}</td>
					<td class="text-right"><strong>${this.fmt_currency(row.amount_to_receive)}</strong></td>
				</tr>
				<tr id="ddr-detail-${frappe.utils.escape_html(dn)}" style="display:none;">
					<td colspan="12" style="background:#fafbfc;padding:0;">
						${this.render_item_drilldown(row)}
					</td>
				</tr>
			`);
		});
	}

	render_item_drilldown(row) {
		const items = row.items || [];
		if (!items.length) {
			return `<div style="padding:12px;" class="text-muted">${__("No book lines")}</div>`;
		}

		const item_rows = items
			.map(
				(it) => `
			<tr>
				<td>${frappe.utils.escape_html(it.item_code || "")}</td>
				<td>${frappe.utils.escape_html(it.item_name || "")}</td>
				<td><span class="label label-default">${frappe.utils.escape_html(it.book_type || "")}</span></td>
				<td>${frappe.utils.escape_html(it.warehouse || "")}</td>
				<td class="text-right">${this.fmt_num(it.qty)}</td>
				<td class="text-right">${this.fmt_num(it.custom_cartons)}</td>
				<td class="text-right">${this.fmt_currency(it.rate)}</td>
				<td class="text-right">${this.fmt_currency(it.amount)}</td>
			</tr>`
			)
			.join("");

		const meta = [
			row.city && `${__("City")}: ${row.city}`,
			row.area && `${__("Area")}: ${row.area}`,
			row.province && `${__("Province")}: ${row.province}`,
			row.courier_mode_of_payment && `${__("Courier Payment")}: ${row.courier_mode_of_payment}`,
			row.invoiced_outstanding > 0 && `${__("Invoiced Outstanding")}: ${this.fmt_currency(row.invoiced_outstanding)}`,
		]
			.filter(Boolean)
			.join(" · ");

		return `
			<div style="padding:12px 16px;">
				<div style="font-size:12px;color:#555;margin-bottom:8px;">${frappe.utils.escape_html(meta)}</div>
				<table class="table table-condensed table-bordered" style="margin:0;background:#fff;">
					<thead>
						<tr style="background:#eef2f6;">
							<th>${__("Item Code")}</th>
							<th>${__("Book / Item")}</th>
							<th>${__("Type")}</th>
							<th>${__("Warehouse")}</th>
							<th class="text-right">${__("Qty")}</th>
							<th class="text-right">${__("Cartons")}</th>
							<th class="text-right">${__("Rate")}</th>
							<th class="text-right">${__("Amount")}</th>
						</tr>
					</thead>
					<tbody>${item_rows}</tbody>
				</table>
			</div>`;
	}

	export_csv() {
		const rows = this.data.rows || [];
		if (!rows.length) {
			frappe.msgprint(__("No data to export"));
			return;
		}

		const lines = [
			[
				"Delivery Note",
				"Date",
				"Customer",
				"Address",
				"City",
				"Area",
				"Warehouse",
				"Delivery Mode",
				"Courier",
				"Item Code",
				"Item Name",
				"Book Type",
				"Item Warehouse",
				"Qty",
				"Cartons",
				"Rate",
				"Amount",
				"Courier Payable",
				"Customer Amount",
				"Amount To Receive",
			].join(","),
		];

		rows.forEach((row) => {
			(row.items || [{}]).forEach((it) => {
				lines.push(
					[
						row.delivery_note_no,
						row.posting_date,
						this.csv_cell(row.customer_name),
						this.csv_cell(row.shipping_address),
						this.csv_cell(row.city),
						this.csv_cell(row.area),
						this.csv_cell(row.warehouses_label),
						this.csv_cell(row.delivery_mode),
						this.csv_cell(row.courier),
						this.csv_cell(it.item_code),
						this.csv_cell(it.item_name),
						this.csv_cell(it.book_type),
						this.csv_cell(it.warehouse),
						it.qty || 0,
						it.custom_cartons || 0,
						it.rate || 0,
						it.amount || 0,
						row.courier_payable || 0,
						row.customer_amount || 0,
						row.amount_to_receive || 0,
					].join(",")
				);
			});
		});

		const blob = new Blob([lines.join("\n")], { type: "text/csv;charset=utf-8;" });
		const url = URL.createObjectURL(blob);
		const a = document.createElement("a");
		a.href = url;
		a.download = `dispatch-detail-${frappe.datetime.get_today()}.csv`;
		a.click();
		URL.revokeObjectURL(url);
	}

	csv_cell(val) {
		const s = String(val == null ? "" : val).replace(/"/g, '""');
		return `"${s}"`;
	}

	fmt_currency(val) {
		return format_currency(flt(val), frappe.defaults.get_default("currency"));
	}

	fmt_num(val) {
		return format_number(flt(val), null, 0);
	}
	};
}
